linked_list/linked_list.py

```diff
@@ -86,32 +170,85 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n^2)
+    # Space Complexity: O(1)
+
+    # This is definitely not the approach  :(
     def reverse(self):
-        pass
-  
+        
+        length = self.length()
+        current = None
+
+        for i in range(length-1):
+            is_last = self.head
+            previous = None
+            while is_last.next != None:
+                previous = is_last
+                is_last = is_last.next
+            if i == 0:
+                is_last.next = self.head
+                self.head = is_last
+                previous.next = None
+                current = is_last
+            else:
+                is_last.next = current.next
+                current.next = is_last
+                previous.next = None
+                current = is_last
+
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        length = self.length()
+
+        current = self.head 
+
+        midpoint = length // 2
+
+        return self.get_at_index(midpoint)
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
     # Time Complexity: ?
-    # Space Complexity: ?
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):

```

Comment:
✨ Super creative solution, I like it! Looks like you missed adding time complexity.